src/csharp/Microsoft.Spark/Sql/Types/DataType.cs

```diff
@@ -188,12 +193,102 @@ private static DataType ParseSimpleType(JToken json)
             throw new ArgumentException($"Could not parse data type: {json}");
         }
 
+        /// <summary>
+        /// Remove "Type" from the end of type name and lower cases to align with Scala type name.
+        /// </summary>
+        /// <param name="type">The type to normalize.</param>
+        /// <returns>Normalized type name.</returns>
+        private static string NormalizeTypeName(Type type)
+        {
+            if (s_simpleTypeNormalizedNames == null)
+            {
+                Debug.Assert(s_complexTypeNormalizedNames == null);
+                BuildNormalizedStringMapping();
+            }
+
+            for (int i = 0; i < s_simpleTypes.Length; i++)
+            {
+                if (s_simpleTypes[i] == type)
+                {
+                    return s_simpleTypeNormalizedNames[i];
+                }
+            }
+
+            for (int i = 0; i < s_complexTypes.Length; i++)
+            {
+                if (s_complexTypes[i] == type)
+                {
+                    return s_complexTypeNormalizedNames[i];
+                }
+            }
+
+            return NormalizeTypeName(type.Name);
+        }
+
         /// <summary>
         /// Remove "Type" from the end of type name and lower cases to align with Scala type name.
         /// </summary>
         /// <param name="typeName">Type name to normalize</param>
         /// <returns>Normalized type name</returns>
-        private static string NormalizeTypeName(string typeName) =>
-            typeName.Substring(0, typeName.Length - 4).ToLower();
+        private static string NormalizeTypeName(string typeName)
+        {
+#if NETSTANDARD2_1
+            return string.Create(typeName.Length - 4, typeName, (span, typeName) =>
+            {
+                typeName.AsSpan(0, typeName.Length - 4).ToLower(span, CultureInfo.CurrentCulture);
+            });
+#else
+            return typeName.Substring(0, typeName.Length - 4).ToLower();
+#endif
+        }
+
+        /// <summary>
+        /// Uses the built up normalized type name cache to find the index of the simple type that matches the passed in <paramref name="typeName"/>.
+        /// </summary>
+        /// <param name="typeName">Normalized simple type name to compare against.</param>
+        /// <returns>The index of the simple type within the list that matches the type name, if found. Otherwise, -1.</returns>
+        private static int SimpleTypeIndex(string typeName)
+        {
+            if (s_simpleTypeNormalizedNames == null)
+            {
+                Debug.Assert(s_complexTypeNormalizedNames == null);
+                BuildNormalizedStringMapping();
+            }
+            return s_simpleTypeNormalizedNames.AsSpan().IndexOf(typeName);
```

Comment:
Why `AsSpan` here? Why not use `Array.IndexOf`?

Reply:
Habit/personal preference to use spans :)

`Array.IndexOf` works fine. Shall I change it?

Reply:
Your call. It was just surprising to get a span just for IndexOf.